File: nand2tetris/jack-compiler/go/vmwriter/vmwriter.go
package vmwriter

import (
	"fmt"
	"os"
)

type Command string

const (
	ADD Command = "add"
	SUB Command = "sub"
	NEG Command = "neg"
	EQ  Command = "eq"
	GT  Command = "gt"
	LT  Command = "lt"
	AND Command = "and"
	OR  Command = "or"
	NOT Command = "not"
)

type Segment string

const (
	STATIC  Segment = "static"
	ARG     Segment = "argument"
	LOCAL   Segment = "local"
	CONST   Segment = "constant"
	THIS    Segment = "this"
	THAT    Segment = "that"
	POINTER Segment = "pointer"
	TEMP    Segment = "temp"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

type VMWriter struct {
	out *os.File
}

func New(pathName string) *VMWriter {
	f, err := os.Create(pathName)
	check(err)

	vm := &VMWriter{out: f}

	return vm
}

func (vm *VMWriter) WritePush(segment Segment, index int) {
	s := fmt.Sprintf("push %s %d\n", segment, index)
	vm.out.WriteString(s)
}

func (vm *VMWriter) WritePop(segment Segment, index int) {
	s := fmt.Sprintf("pop %s %d\n", segment, index)
	vm.out.WriteString(s)
}

func (vm *VMWriter) WriteArithmetic(command Command) {
	s := fmt.Sprintf("%s\n", command)
	vm.out.WriteString(s)
}

func (vm *VMWriter) WriteCall(name string, nArgs int) {
	s := fmt.Sprintf("call %s %d\n", name, nArgs)
	vm.out.WriteString(s)
}

func (vm *VMWriter) WriteFunction(name string, nLocals int) {
	s := fmt.Sprintf("function %s %d\n", name, nLocals)
	vm.out.WriteString(s)
}

func (vm *VMWriter) WriteReturn() {
	s := fmt.Sprintf("return\n")
	vm.out.WriteString(s)
}

func (vm *VMWriter) WriteLabel(label string) {
	s := fmt.Sprintf("label %s\n", label)
	vm.out.WriteString(s)
}

func (vm *VMWriter) WriteGoto(label string) {
	s := fmt.Sprintf("goto %s\n", label)
	vm.out.WriteString(s)
}

func (vm *VMWriter) WriteIf(label string) {
	s := fmt.Sprintf("if-goto %s\n", label)
	vm.out.WriteString(s)
}

func (vm *VMWriter) CloseFile() {
	vm.out.Close()
}

File: nand2tetris/jack-compiler/cpp/vmwriter.cpp
#include "vmwriter.h"
#include <fstream>
#include <iostream>
#include <string>
#include <map>
using namespace std;

std::map<Segment, std::string>
    segments = {
        {CONST, "constant"},
        {ARG, "argument"},
        {LOCAL, "local"},
        {THIS, "this"},
        {THAT, "that"},
        {POINTER, "pointer"},
        {STATIC, "static"},
        {TEMP, "temp"},

};

std::map<Command, std::string>
    commands = {
        {ADD, "add"},
        {SUB, "sub"},
        {NEG, "neg"},
        {EQ, "eq"},
        {GT, "gt"},
        {LT, "lt"},
        {AND, "and"},
        {OR, "or"},
        {NOT, "not"},

};

VMWriter::VMWriter(string fileName) { out.open(fileName.c_str()); }

void VMWriter::close(void)
{
  if (out.is_open())
    out.close();
}

void VMWriter::writePush(Segment segment, int index)
{
  out << "push " << segments[segment] << " " << index << endl;
}

void VMWriter::writePop(Segment segment, int index)
{
  out << "pop " << segments[segment] << " " << index << endl;
}

void VMWriter::writeArithmetic(Command command)
{
  out << commands[command] << endl;
}

void VMWriter::writeLabel(string label) { out << "label " << label << endl; }
void VMWriter::writeGoto(string label) { out << "goto " << label << endl; }

void VMWriter::writeIf(string label) { out << "if-goto " << label << endl; }

void VMWriter::writeCall(string name, int nArgs)
{
  out << "call " << name << " " << nArgs << endl;
}

void VMWriter::writeFunction(string name, int nLocals)
{
  out << "function " << name << " " << nLocals << endl;
}

void VMWriter::writeReturn(void) { out << "return" << endl; }